fuel-merkle/src/sparse/proof.rs

```diff
@@ -0,0 +1,692 @@
+use crate::{
+    common::{
+        path::{
+            Instruction,
+            Path,
+        },
+        Bytes32,
+        Prefix,
+        ProofSet,
+    },
+    sparse::Node,
+};
+
+use alloc::vec::Vec;
+use core::{
+    fmt,
+    fmt::Debug,
+};
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub enum Proof {
+    Inclusion(InclusionProof),
+    Exclusion(ExclusionProof),
+}
+
+impl Proof {
+    pub fn root(&self) -> &Bytes32 {
+        match self {
+            Proof::Inclusion(proof) => &proof.root,
+            Proof::Exclusion(proof) => &proof.root,
+        }
+    }
+
+    pub fn proof_set(&self) -> &ProofSet {
+        match self {
+            Proof::Inclusion(proof) => &proof.proof_set,
+            Proof::Exclusion(proof) => &proof.proof_set,
+        }
+    }
+
+    pub fn is_inclusion(&self) -> bool {
+        match self {
+            Proof::Inclusion(_) => true,
+            Proof::Exclusion(_) => false,
+        }
+    }
+
+    pub fn is_exclusion(&self) -> bool {
+        !self.is_inclusion()
+    }
+}
+
+#[derive(Clone, Eq, PartialEq)]
+pub struct InclusionProof {
+    pub root: Bytes32,
+    pub proof_set: ProofSet,
+}
+
+impl InclusionProof {
+    pub fn verify<K: Into<Bytes32>, V: AsRef<[u8]>>(&self, key: K, value: &V) -> bool {
```

Comment:
You need to use `MerkleTreeKey` as an input `key` type or hash it under the hood.